[2단계 - 자동차 경주 리팩터링] 이상(이건상) 미션 제출합니다.

kunsanglee opened the pull request:
안녕하세요 알렉스! 😄
2단계에서 해야되는 것은 두 가지입니다.
> 1. 핵심 비지니스 로직을 가지는 객체를 domain 패키지, UI 관련한 객체를 view 패키지에 구현한다.
> 2. MVC 패턴 기반으로 리팩터링해 view 패키지의 객체가 domain 패키지 객체에 의존할 수 있지만, domain 패키지의 객체는 view 패키지 객체에 의존하지 않도록 구현한다.

두 번째 항목은 반영되어 있어서 Main 클래스의 이름을 RacingCarMain으로 변경하고 
InputView와 OutputView 클래스를 view 패키지로 분리하였습니다.

2단계의 문서 중 MVC 패턴으로 변경 후의 main 메소드 예시 코드를 보다가 궁금한 것이 생겼는데요

```java
import view.InputView;
import view.ResultView;
import domain.RacingGame;

public class RacingMain {
    public static void main(final String... args) {
        final var carNames = InputView.getCarNames();
        final var tryCount = InputView.getTryCount();

        final var racingGame = new RacingGame(carNames, tryCount);
        racingGame.race();
        
        ResultView.printWinners(racingGame.getWinners());
    }
}

```

final var 키워드를 사용한 타입 추론 기능으로 패키지 의존성을 제거하는 효과를 확인했는데, 사실상 가독성이 그렇게까지 좋다고 느끼지 못했습니다. 원래 제 코드에서는 InputView의 메서드 결과를 반환받아서 명확한 타입 Cars, Round 타입의 변수로 할당하여 import 구문이 추가되었습니다.

```java
package racinggame;

import racinggame.domain.Cars; // Cars 의존 추가
import racinggame.domain.RacingGame;
import racinggame.domain.Round; // Round 의존 추가
import racinggame.domain.condition.RandomMoveCondition;
import racinggame.view.InputView;
import racinggame.view.OutputView;

public class RacingCarMain {

    public static void main(String[] args) {
        Cars cars = repeat(InputView::readCars);
        Round round = repeat(InputView::readRound);
        RacingGame game = RacingGame.of(cars, new RandomMoveCondition());

        OutputView.printResult(game.race(round));
        OutputView.printWinner(game.findWinnerName());
    }
}
```

여기서 변수에 할당하는 대신 바로 메서드의 인자로 넣는 두 번째 코드를 통해 import 구문을 없앨 수 있었는데요.
1번 케이스인 var 키워드를 사용하는 것, 
2번 케이스인 명시적 타입 변수 할당, 
3번 케이스인 메서드의 인자로 바로 할당하는 것 
이 세 가지 케이스 중 어떤 것이 가장 실무에서 많이 사용하는지 궁금합니다! 
물론 아직 패키지 의존성까지 신경써야 할 수준은 아닌 것은 알고있으나 궁금하여 여쭤봅니다 🥲 

```java
package racinggame;

// Cars, Round import 제거됨.
import racinggame.domain.RacingGame;
import racinggame.domain.condition.RandomMoveCondition;
import racinggame.view.InputView;
import racinggame.view.OutputView;

public class RacingCarMain {

    public static void main(String[] args) {
        RacingGame game = RacingGame.of(repeat(InputView::readCars), new RandomMoveCondition());

        OutputView.printResult(game.race(repeat(InputView::readRound)));
        OutputView.printWinner(game.findWinnerName());
    }
}
```

kunsanglee committed:
refactor : Name 생성시 검증 메서드에서 발생하는 에러 메세지에 선언한 상수 동기화

kunsanglee committed:
refactor : Position 객체를 음수로 생성시 구체적 에러를 던지도록 수정

kunsanglee committed:
refactor : Main 클래스 이름 변경 및 InputView, OutputView view 패키지로 이동

yxxnghwan commented on src/main/java/racinggame/RacingCarMain.java:
```diff
@@ -1,20 +1,17 @@
-        Round round = repeat(InputView::readRound);
-        RacingGame game = RacingGame.of(cars, new RandomMoveCondition());
-
-        OutputView.printResult(game.race(round));
+    public static void main(String[] args) {
+        RacingGame game = RacingGame.of(repeat(InputView::readCars), new RandomMoveCondition());
+        
+        OutputView.printResult(game.race(repeat(InputView::readRound)));
```
음 한 라인에 너무 많은 일들을 수행하고 있는 것 같아요! 이전이 좀 더 나아보이는데, 이상의 생각은 어떠신가요?

yxxnghwan commented on src/main/java/racinggame/domain/Name.java:
```diff
@@ -27,7 +27,8 @@ private static void checkNameIsEmpty(String name) {
 
     private static void checkNameLength(String name) {
         if (name.length() > MAX_NAME_LENGTH) {
-            throw new IllegalArgumentException(String.format("자동차 이름은 최대 5자 이하만 가능합니다. 입력한 자동차 이름: %s", name));
+            throw new IllegalArgumentException(
+                    String.format("자동차 이름은 최대 %d자 이하만 가능합니다. 입력한 자동차 이름: %s", MAX_NAME_LENGTH, name));
```
👍🏿

yxxnghwan commented on src/main/java/racinggame/domain/Position.java:
```diff
@@ -22,7 +22,7 @@ public static Position from(int position) {
 
     private static void checkIsNonNegative(int position) {
         if (position < ZERO) {
-            throw new RuntimeException();
+            throw new IllegalArgumentException(String.format("자동차의 위치는 양수만 가능합니다. 입력한 값: %s", position));
```
에러 변경해주셨군요! 👍🏿

yxxnghwan commented on src/main/java/racinggame/RacingCarMain.java:
```diff
@@ -1,20 +1,17 @@
-
-        OutputView.printResult(game.race(round));
+    public static void main(String[] args) {
+        RacingGame game = RacingGame.of(repeat(InputView::readCars), new RandomMoveCondition());
```
> 이전 단계에서 리팩터링중에 RacingGame의 of 메서드 안에서 수행하던 검증 로직을 Cars 클래스로 옮기면서
> 단순히 RacingGame 생성자를 호출하고 반환하는 로직만 남게 되었습니다.
> 
> CarInfo, RoundResult와 같은 DTO 성격이 아닌 도메인 클래스에 이미 from, of 와 같은 new 키워드를 통한 생성을 막고 팩토리 메서드를 사용하여 생성하고 있어서 생성시 통일성을 위해 삭제하지 않았는데요
> 이런 상황에는 과감히 RacingGame 클래스에는 팩토리 메서드를 삭제하는 것이 좋을지 고민이 됩니다!
> 
> 알렉스는 어떻게 생각하시나요?

이상이 생각하시는 정적 팩토리 메소드의 역할은 무엇인가요? 개인적으로 정적 팩토리 메소드에 기대하는 부분은 단순한 생성자 포팅은 아니라고 생각해요! 현재 `RacingGame.of()` 메소드에선 파라미터를 그대로 `new RacingGame()` 포팅해서 전달하고 있기 때문에 불필요한 메소드를 정의한 것처럼 느껴집니다!

yxxnghwan commented on src/main/java/racinggame/view/InputView.java:
```diff
@@ -9,7 +9,7 @@
 
-class InputView {
+public class InputView {
 
```
입력을 담당하는 클래스에서 `Cars` 가 만들어지고 있는데, Cars를 만드는 로직은 비즈니스 로직이 아닐까요?

yxxnghwan reviewed (changes requested):
안녕하세요 이상, 기존 1단계에서 MVC 구조로 작성해주셔서 2단계 코드 수정이 많이 없었군요..!👍🏿
이전 PR에서 못다한 답변과 새 코멘드 추가했습니다! 질문에 대한 답변도 아래에 남겼으니 같이 고민해보아요!🚀

> 여기서 변수에 할당하는 대신 바로 메서드의 인자로 넣는 두 번째 코드를 통해 import 구문을 없앨 수 있었는데요.
> 
> 1번 케이스인 var 키워드를 사용하는 것,
> 
> 2번 케이스인 명시적 타입 변수 할당,
> 
> 3번 케이스인 메서드의 인자로 바로 할당하는 것
> 
> 이 세 가지 케이스 중 어떤 것이 가장 실무에서 많이 사용하는지 궁금합니다!
> 
> 물론 아직 패키지 의존성까지 신경써야 할 수준은 아닌 것은 알고있으나 궁금하여 여쭤봅니다 🥲

실무에서 자주 사용하는 방법은 따로 정해져있진 않습니다! 상황에 따라 다르게 쓰기도 하고 팀 컨벤션에 따라 좌지우지 되기도 하고, 개인 취향에 따라 갈리기도 하는 부분이에요! 정답이 없다보니..

의존성에 대해서는 좀 더 고민해보시면 좋을 것 같아요! import가 추가되지 않는다고 의존성이 정말 생기지 않은게 맞을지..! 이상이 생각하는 의존성의 기준도 얘기해주시면 재밌을 것 같네요..! ㅎㅎ

개인적으로 이번 코드 변경을 보고서는 기존의 2번 방식이 가독성이 가장 좋았습니다! 이유도 코멘트로 달아두었으니 확인 부탁드려요!

kunsanglee committed:
refactor : 코드 리팩토링

- Cars 클래스의 팩토리 메서드 of() 에서 List<Name> 타입을 매개변수로 받아서 자동차들을 생성하는 로직으로 변경했다.
- Name 클래스의 접근제어자를 default에서 public으로 수정하여 RacingCarMain 클래스에서 InputView를 통해 받는 것으로 변경했다.
- RacingGame 클래스에서 의미없이 사용하던 of() 메서드를 삭제하고 생성자의 접근 제어자를 public 으로 변경하여 외부에서 생성하도록 변경했다.
- InputView readNames() 메서드에서 Cars를 생성하지 않고 List<Name>을 반환하도록 변경했다.

kunsanglee committed:
refactor : InputView readNames() 메서드 수정

yxxnghwan commented on src/main/java/racinggame/view/InputView.java:
```diff
@@ -1,28 +1,37 @@
+public class InputView {
 
     private static final String DELIMITER = ",";
+    private static final int MIN_ROUND = 1;
```
사용하지 않는 상수인 것 같아요!

yxxnghwan commented on src/main/java/racinggame/view/InputView.java:
```diff
@@ -1,28 +1,37 @@
+    }
+
+    private static void checkCarNameSize(List<Name> names) {
+        if (names.size() < 2) {
```
2라는 매직넘버도 상수화 되어 의미를 표현해줬으면 좋았을 것 같습니다!

yxxnghwan reviewed (approved):
첫 미션 수고하셨습니다 이상! 간단한 코멘트만 남겼고 다음 미션 진행하시면서 유의해서 진행해주시면 좋을 것 같습니다! 이번 미션에서는 충분히 잘 해주셔서 이만 머지하도록 할게요! 앞으로의 우테코 생활도 화이팅입니다! 💪
